### PR TITLE
Submission: Top-Heavy FFN Allocation + Packed Int6 Export | pending eval

### PR DESCRIPTION
**Pull request by mr-ashish-panday**
### Approach

This submission focuses on **parameter allocation**, not just export format. The current public frontier is clustered around a uniform `9 x 512` backbone with stronger quantization and evaluation tricks. I kept the proven surrounding recipe, but replaced the uniform `3x` FFN with **OpenELM-style layer-wise scaling** so later blocks get larger FFNs and earlier blocks get smaller ones at the same overall budget.

The exact FFN schedule is:

`768, 960, 1152, 1344, 1536, 1728, 1920, 2112, 2304`

The motivation is that, under a strict parameter budget, later layers are closer to the next-token loss and should benefit more from extra capacity than early layers. This is especially attractive in the Parameter Golf setting because the challenge is strongly **parameter-limited** relative to the 10-minute 8xH100 compute budget. Chinchilla-style reasoning says the available training compute could support a much larger model than the 16 MB artifact allows, so the right move is to spend the byte budget as efficiently as possible inside the model rather than leaving it uniformly distributed.

To support that architecture cleanly, I also implemented an **exact packed-int6 export path** with per-row fp16 scales. This keeps the submission self-contained and avoids relying on an external `zstd` dependency at evaluation time. I keep the tied embedding in fp16 because that tensor is especially quantization-sensitive and serves as both the input embedding and output projection.

Finally, the trainer includes **sliding-window evaluation** and a **CPU `DRY_RUN=1` mode** so the full train/export/eval pipeline can be verified locally without CUDA. Because I do not currently have GPU credits, I am not claiming a score yet and am leaving final eval metrics pending.

### Key Design Decisions

- Use **top-heavy FFN allocation** instead of a uniform `3x` FFN so later layers get more of the parameter budget.
- Keep the proven **9-layer, 512-dim, tied-embedding, GQA** backbone for a controlled comparison against public baselines.
- Export large matrices as **packed int6 + per-row fp16 scale + zlib** to make the artifact fully self-contained.
- Keep `tok_emb.weight` in **fp16** because it is the most quantization-sensitive tensor in public submissions.
- Retain **sliding-window eval** because it is allowed by the rules and is consistently helpful in public runs.
- Use **seed 42** by default and add a **10-step CPU dry-run** path for reproducibility without GPUs.

### What I Expect to Underperform / Risks

- The main risk is that layer-wise FFN scaling is theoretically motivated but not yet validated on this exact challenge objective.
- Packed int6 + zlib is self-contained, but the trained checkpoint still needs to prove it compresses comfortably under the 16 MB cap.
- If this challenge is dominated more by export/eval tricks than by parameter placement, the gain over a strong uniform-FFN baseline may be modest.

### Results

| Metric | Value |
|---|---|
| Baseline eval loss | `1.22436570` val_bpb (repo baseline) |
| Final eval loss | Pending — training to be run once compute credits are granted |
| Parameter count | `21,778,504` |
| Artifact size | Measured init-export: `4,273,390` bytes total; dense-random stress probe: `16,549,133` bytes total |
| Training time (estimated on 8×H100) | Expected to fit the 600s budget; public 21.8M-parameter 9x512 runs report roughly `45–57 ms/step`, implying about `10k–13k` steps in 10 minutes |

### Reproducibility

Seed: `42`.

Dry-run verified locally with:

`DRY_RUN=1 RUN_ID=topheavy_dryrun python train_gpt.py`

Full 8xH100 train/eval pending compute access.

### References

- Hoffmann et al., *Training Compute-Optimal Large Language Models*, 2022
- Mehta et al., *OpenELM*, 2024

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Submission: Top-Heavy FFN Allocation + Packed Int6 Export | pending eval

**BPB:** (not parsed — see PR title) | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `411d753b1c4e`, file `records/track_10min_16mb/2026-03-19_TopHeavyFFN_PackedInt6_PendingEval/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=1024, code=60837 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=1024, code=60837 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
